src/lib/x509/certstor.cpp

```diff
@@ -213,4 +214,135 @@ Certificate_Store_In_Memory::Certificate_Store_In_Memory(const std::string& dir)
+
+   public:
+      const std::string m_system_store;
+      std::vector<std::shared_ptr<const X509_Certificate>> m_certificates;
+   };
+
+Flatfile_Certificate_Store::Flatfile_Certificate_Store() :
+   m_impl(std::make_shared<Flatfile_Certificate_Store_Impl>(""))
+   {
+   }
+
+#if defined(BOTAN_TARGET_OS_HAS_FILESYSTEM)
+Flatfile_Certificate_Store::Flatfile_Certificate_Store(const std::string& file) :
+   m_impl(std::make_shared<Flatfile_Certificate_Store_Impl>(file))
+   {
+   }
+#endif
+
+std::vector<X509_DN> Flatfile_Certificate_Store::all_subjects() const
+   {
+   std::vector<X509_DN> dns;
+   std::transform(m_impl->m_certificates.cbegin(), m_impl->m_certificates.cend(),
+   std::back_inserter(dns), [](std::shared_ptr<const X509_Certificate> cert) { return cert->subject_dn();});
+
+   return  dns;
+   }
+
+std::shared_ptr<const X509_Certificate>
+Flatfile_Certificate_Store::find_cert(const X509_DN& subject_dn,
+                                   const std::vector<uint8_t>& key_id) const
+   {
+   const auto found_cert = std::find_if(m_impl->m_certificates.cbegin(), m_impl->m_certificates.cend(),
+                                        [&](std::shared_ptr<const X509_Certificate> cert)
+      {
+      if(!key_id.empty() && !cert->subject_key_id().empty())
+         {
+         return key_id == cert->subject_key_id() &&
+                subject_dn == cert->subject_dn();
+         }
+      return subject_dn == cert->subject_dn();
+      });
+
+   if(found_cert != m_impl->m_certificates.cend())
+      {
+      return *found_cert;
+      }
+
+   return nullptr;
+   }
+
+std::vector<std::shared_ptr<const X509_Certificate>> Flatfile_Certificate_Store::find_all_certs(
+         const X509_DN& subject_dn,
+         const std::vector<uint8_t>& key_id) const
+   {
+   std::vector<std::shared_ptr<const X509_Certificate>> filtered_certificates;
+
+   std::copy_if(m_impl->m_certificates.cbegin(), m_impl->m_certificates.cend(), std::back_inserter(filtered_certificates),
+                [&](std::shared_ptr<const X509_Certificate> cert)
+      {
+      if(!key_id.empty() && !cert->subject_key_id().empty())
+         {
+         return key_id == cert->subject_key_id() &&
+                subject_dn == cert->subject_dn();
+         }
+      return subject_dn == cert->subject_dn();
+      });
+
+   return filtered_certificates;
+   }
+
+std::shared_ptr<const X509_Certificate>
+Flatfile_Certificate_Store::find_cert_by_pubkey_sha1(const std::vector<uint8_t>& key_hash) const
+   {
+   if(key_hash.size() != 20)
+      {
+      throw Invalid_Argument("Flatfile_Certificate_Store::find_cert_by_pubkey_sha1 invalid hash");
+      }
+
+   std::unique_ptr<HashFunction> hash(HashFunction::create("SHA-1"));
+
+   const auto found_cert = std::find_if(m_impl->m_certificates.cbegin(), m_impl->m_certificates.cend(),
+                                        [&](std::shared_ptr<const X509_Certificate> cert)
+      {
+      hash->update(cert->subject_public_key_bitstring());
```

Comment:
This hash is precomputed when certificates are parsed, use `X509_Certificate::subject_public_key_bitstring_sha1`